File: rust/src/day9.rs
pub struct Submarine {
    heightmap: Vec<Vec<u32>>,
    visited: Vec<Vec<bool>>,
}

impl Submarine {
    pub fn new() -> Self {
        Self {
            heightmap: Vec::new(),
            visited: Vec::new(),
        }
    }

    pub fn on_input(&mut self, result: &str) {
        self.heightmap.push(result.chars().map(|c| c.to_digit(10).unwrap()).collect());
        self.visited.push(vec![false; self.heightmap[0].len()]);
    }

    fn low_points(&self) -> Vec<(usize, usize, u32)> {
        let mut result = Vec::new();

        for i in 0..self.heightmap.len() {
            let row = &self.heightmap[i];

            for j in 0..row.len() {
                let height = row[j];

                // If any of the conditions below are true, the point is not a low point, and we won't get to the sum below.
                if j > 0 && row[j - 1] <= height {
                    continue;
                }
                if j < row.len() - 1 && row[j + 1] <= height {
                    continue;
                }
                if i > 0 && self.heightmap[i - 1][j] <= height {
                    continue;
                }
                if i < self.heightmap.len() - 1 && self.heightmap[i + 1][j] <= height {
                    continue;
                }

                result.push((i, j, height));
            }
        }

        result
    }

    fn find_basin_size(&mut self, (y, x): (usize, usize)) -> u32 {
        if self.visited[y][x] {
            return 0;
        }

        if self.heightmap[y][x] == 9 {
            return 0;
        }

        self.visited[y][x] = true;

        let left = if x == 0 {
            0
        } else {
            self.find_basin_size((y, x - 1))
        };

        let up = if y == 0 {
            0
        } else {
            self.find_basin_size((y - 1, x))
        };

        let right = if x >= self.heightmap[0].len() - 1 {
            0
        } else {
            self.find_basin_size((y, x + 1))
        };

        let down = if y >= self.heightmap.len() - 1 {
            0
        } else {
            self.find_basin_size((y + 1, x))
        };

        1 + left + right + up + down
    }

    pub fn output(&mut self) {
        let low_points = self.low_points();

        println!("Part 1: {}", low_points.iter().map(|x| x.2 + 1).sum::<u32>());

        let mut basin_sizes: Vec<u32> = self.low_points().iter().map(|x| self.find_basin_size((x.0, x.1))).collect();
        basin_sizes.sort_unstable_by(|a, b| b.cmp(a));

        println!("Part 2: {}", basin_sizes.iter().take(3).product::<u32>());
    }
}

File: 2021/julia/Day9.jl
module Day9

export Submarine, on_input!, output

mutable struct Submarine
    heightmap::Matrix{UInt32}
    visited::Matrix{Bool}

    Submarine() = new(Matrix(undef, 0, 0), Matrix(undef, 0, 0))
end

function on_input!(s::Submarine, a::AbstractString)
    heights = parse.(UInt32, split(a, ""))

    if size(s.heightmap) == (0, 0)
        s.heightmap = Matrix(undef, 0, length(heights))
        s.visited = Matrix(undef, 0, length(heights))
    end

    s.heightmap = [s.heightmap; heights']
    s.visited = [s.visited; falses(length(heights))']
end

function find_low_points(s::Submarine)
    result = []

    (ysize, xsize) = size(s.heightmap)

    for y in 1:ysize
        for x in 1:xsize
            height = s.heightmap[y, x]

            if x > 1 && s.heightmap[y, x - 1] <= height
                continue
            end
            if x < xsize && s.heightmap[y, x + 1] <= height
                continue
            end
            if y > 1 && s.heightmap[y - 1, x] <= height
                continue
            end
            if y < ysize && s.heightmap[y + 1, x] <= height
                continue
            end

            push!(result, (y, x, height))
        end
    end

    result
end

function find_basin_size!(s::Submarine, (y, x))
    if s.visited[y, x]
        return 0
    end

    if s.heightmap[y, x] == 9
        return 0
    end

    (ysize, xsize) = size(s.heightmap)

    s.visited[y, x] = true

    left = x == 1 ? 0 : find_basin_size!(s, (y, x - 1))
    right = x == xsize ? 0 : find_basin_size!(s, (y, x + 1))
    up = y == 1 ? 0 : find_basin_size!(s, (y - 1, x))
    down = y == ysize ? 0 : find_basin_size!(s, (y + 1, x))

    1 + left + right + up + down
end

function output(s::Submarine)
    low_points = find_low_points(s)
    low_point_risk_level = sum(map(x -> x[3] + 1, low_points))

    println("Part 1: $(low_point_risk_level)")

    basin_sizes = sort(map(x -> find_basin_size!(s, (x[1], x[2])), low_points), rev=true)
    top_3_basin_product = prod(Iterators.take(basin_sizes, 3))

    println("Part 2: $(top_3_basin_product)")
end

end